### PR TITLE
[GraphBolt] enable indexing on ItemSet instance

### PR DESCRIPTION
**Pull request by Rhett-Ying**
## Description

This PR is part of performance improvement for shuffle/batching in `ItemSampler`.

see more details in https://quip-amazon.com/bIfOAW6GMsIG/Benchmark-on-rgcnhetero#temp:C:FIGba38038eb04347578db86f3f1
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/graphbolt/itemset.py [hunk 1]
```diff
@@ -76,22 +89,8 @@ def __init__(
...
-                    self._names = (names,)
-                assert (
-                    len(self._names) == 1
-                ), "Number of names mustn't exceed 1 when item is an integer."
-            else:
-                self._names = None
-            return
-
-        # Otherwise.
+            items = torch.arange(items)
```
[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply python/dgl/graphbolt/itemset.py:93]: yes. the previous impl is lazy instantiated in this case. For now, I move it to __init__

[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply python/dgl/graphbolt/itemset.py:93]: I think it's reasonable to allocate memory when instantiating `ItemSet`. As for the lazy mechanism, I think we should apply into `all_nodes_set()` of `OnDiskDataset`.

[the remark this answers is not in the document]
Comment (frozenbugs) [reply python/dgl/graphbolt/itemset.py:93]: This is probably very space in-efficient. What about just initiate it with a int, but do some magic in __getitem__ method?

[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply python/dgl/graphbolt/itemset.py]: updated with a single defer to avoid degrading `__iter__`.

**Review by frozenbugs — APPROVED** [review]

**Pull request merged**